Source/CombatExtended/Harmony/Compatibility/Harmony_SaveOurShip.cs

```diff
@@ -0,0 +1,80 @@
+using HarmonyLib;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using Verse;
+using System.Reflection.Emit;
+using System;
+using RimWorld;
+using Verse.AI;
+
+namespace CombatExtended.HarmonyCE.Compatibility
+{
+    [HarmonyPatch]
+    class Harmony_Compatibility_SaveOurShip
+    {
+        static readonly Assembly ass = AppDomain.CurrentDomain.GetAssemblies().
+                                        SingleOrDefault(assembly => assembly.GetName().Name == "ShipsHaveInsides");
+        static MethodBase targetMethod = null;
+
+	internal static bool Prepare()
+	{
+	    if (ass == null)
+	    {
+		return false;
+	    }
+	    foreach (var t in ass.GetTypes())
+	    {
+		if (t.Name == "WorkGiver_LoadTorpedoTube")
+		{
+		    targetMethod = AccessTools.Method(t, "FindAmmo");
+		}
+	    }
+	    if (targetMethod == null)
+	    {
+		Log.Error($"Failed to find target method while attempting to patch SaveOurShip.");
+		return false;
+	    }
+	    return true;
+	}
+
+	internal static MethodBase TargetMethod()
+        {
+            return targetMethod;
+        }
+
+	public static void Postfix(Pawn pawn, Building_Turret tube, ref Thing __result)
+	{
+	    if (__result!=null) {
+		return;
+	    }
+	    var turret = tube as Building_ShipTurret;
+	    StorageSettings allowedShellsSettings = ThingCompUtility.TryGetComp<CompChangeableProjectilePlural>(turret.gun).allowedShellsSettings;
+	    Predicate<Thing> pred = (Thing x) =>
+            {
+	        if (!ForbidUtility.IsForbidden(x, pawn) && ReservationUtility.CanReserve(pawn, x, 1, -1, (ReservationLayerDef)null, false))
+                {
+                    if (allowedShellsSettings != null)
+                    {
+                        return allowedShellsSettings.AllowedToAccept(x);
+                    }
+                    return true;
+                }
+                return false;
+            };
```

Comment:
Prefer early returns:

```suggestion
	        if (ForbidUtility.IsForbidden(x, pawn) || !ReservationUtility.CanReserve(pawn, x, 1, -1, (ReservationLayerDef)null, false))
                {
                  return false;
                }
                return allowedShellSettings?.AllowedToAccept(x) ?? false;
             };
```

Reply:
Should be `?? true;` to match the method in SOS, but otherwise looks good.  I suspect launchers not owned by the player may not have an `allowedShellSettings` (which is *not* why this failed for player pawns, the Predicate never gets invoked in the original method).